doc/neps/nep-0029-dtype-as-type.rst

```diff
@@ -0,0 +1,268 @@
+
+    class Colors(UInt8Descr):
+        colors = ['red', 'green', 'blue']
+        def format(value):
+            return Colors.colors[value]
+        ArrFuncs = null_arrayfuncs
+
+    c = np.array([0, 1, 1, 0, 2], dtype=Colors)    
+
+Additional code would be needed to neutralize the `tp_as_number` slot functions.
+
+Advantages
+==========
+
+It is very difficult today to override dtype behaviour, since internally
+descriptor objects are not true type instances, rather contianers for the
+``ArrayDescrObject`` struct.
+
+Disadvantages
+=============
+
+Making descriptors into type objects requires thinking about type classes,
+which is more difficult to reason about than object instances. For instance,
+note that in the ``Colors`` example, we did not instantiate an object of the
+``Colors`` type, rather used that type directly in the ndarray creation. Also
+the ``format`` function is not a bound method of a class instance, rather an
+unbound function on a type class (no ``self`` argument is used).
+
+Future Extensions
+=================
+
+Note the descriptor holds a parallel ``typeobj`` which is a scalar class. A
+call like ``np.dtype('int8')(10)`` will now create a scalar object. The next
+step will be to replace the scalar classes with the descriptor classes, so
+that looking up a scalar's corresponding descriptor type becomes ``type(scalar)``.
+
+We could refactor `numpy.datetime64` to use the new heirarchy, inheriting from
+``np.dtype(uint64)``
+
+Alternatives
+============
+
+Descriptors as Instances
+------------------------
+
+It is confusing that descriptors are classes, not class instances. We could
+define them slightly differently as instances (note the call in the value of
+``dtype_int_dict`` and that ``_repr__`` is now a bound class method of
+``IntDescr``::
+
+    import numpy as np
+
+    class Dtype(type):
+
+        def __new__(cls, obj, *args, **kwargs):
+            if isinstance(obj, int):
+                return dtype_int_dict[obj]
+            elif isinstance(obj, type) and issubclass(obj, np.generic):
+                return dtype_scalar_dict[obj]
+            elif len(args) < 1:
+                # Dtype('int8') or Dtype('S10') or record descr
+                return create_new_descr(cls, obj, *args, **kwargs)
+            else:
+                return super().__new__(cls, obj, *args, **kwargs)
+
+        def __call__(self, args, kwargs):
+            return super().__call__(self.__name__, args, kwargs)
+
+    class GenericDescr(type, metaclass=Dtype):
+        def __new__(cls, *args, **kwargs):
+            import pdb;pdb.set_trace()
+            return type.__new__(cls, *args, **kwargs)
+
+        def __call__(self, *args, **kwargs):
+            return self.typeobj(*args, **kwargs)
+
+    class IntDescr(GenericDescr):
+        def format(value):
+            return '%d' % value
+        def __repr__(self):
+            return 'dtype(%s%d)' %(self.kind, self.itemsize)
+
+
+    class UInt8Descr(IntDescr):
```

Comment:
In this example, wouldn't `uint8descr` be an instance of `IntDescr`, i.e., `uint8descr = IntDescr(<something>)`?